src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs

```diff
@@ -211,22 +211,31 @@ protected override Expression VisitConstant(ConstantExpression c)
 
 
         private static ILookup<string, MethodInfo>? s_seqMethods;
-        private static MethodInfo FindEnumerableMethod(string name, ReadOnlyCollection<Expression> args, params Type[]? typeArgs)
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2060:MakeGenericMethod",
+            Justification = "Enumerable methods don't have trim annotations.")]
+        private static MethodInfo FindEnumerableMethodForQueryable(string name, ReadOnlyCollection<Expression> args, params Type[]? typeArgs)
         {
             if (s_seqMethods == null)
             {
-                s_seqMethods = typeof(Enumerable).GetStaticMethods().ToLookup(m => m.Name);
+                s_seqMethods = GetEnumerableStaticMethods(typeof(Enumerable)).ToLookup(m => m.Name);
             }
             MethodInfo? mi = s_seqMethods[name].FirstOrDefault(m => ArgsMatch(m, args, typeArgs));
             Debug.Assert(mi != null, "All static methods with arguments on Queryable have equivalents on Enumerable.");
             if (typeArgs != null)
                 return mi.MakeGenericMethod(typeArgs);
             return mi;
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
+                Justification = "This method is intentionally hiding the Enumerable type from the trimmer so it doesn't preserve all Enumerable's methods. " +
+                "This is safe because all Queryable methods have a DynamicDependency to the corresponding Enumerable method.")]
+            static MethodInfo[] GetEnumerableStaticMethods(Type type) =>
+                type.GetMethods(BindingFlags.Public | BindingFlags.Static);
         }
 
+        [RequiresUnreferencedCode("Enumerating in-memory collections as IQueryable can require unreferenced code because expressions referencing IQueryable extension methods can get rebound to IEnumerable extension methods. The IEnumerable extension methods could be trimmed causing the application to fail at runtime.")]
         private static MethodInfo FindMethod(Type type, string name, ReadOnlyCollection<Expression> args, Type[]? typeArgs)
```

Comment:
It's not really possible to bubble it up the "normal" way to a public API.

EnumerableRewriter gets created in 2 places:

https://github.com/dotnet/runtime/blob/56d0eb94ec83473d99c1050072a224da327bc02a/src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs#L97-L102

https://github.com/dotnet/runtime/blob/56d0eb94ec83473d99c1050072a224da327bc02a/src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableExecutor.cs#L33-L36

The "normal" way of bubbling it up would mean that:

1. All calls to `GetEnumerator()` would need to be marked as `RequiresUnreferencedCode`, because this `IEnumerable` is unsafe.
2. Similarly all `ExpressionVisitor`s would be marked as `RequiresUnreferencedCode`, because one derived class has an unsafe override.
3. The same reasoning would be applied to the `IQueryProvider.Execute` methods.

If we "jump" over to what could possibly cause these methods to be invoked, it really only comes from one place: `EnumerableQuery<T>`. We could mark its constructor as `RequiresUnreferencedCode`. This would mean calling `IEnumerable<T>.AsQueryable()` would be `RequiresUnreferencedCode` and cause warnings whenever it is called. This is kind of unfortunate because:

1. Creating the queryable isn't unsafe.
2. Typically, enumerating the queryable isn't unsafe.
3. What's really only unsafe is creating your own extension methods on `IQueryable`, and then using them against in-memory collections.

For 90%+ of developers who use these APIs, this is just going to be noise. I didn't even know about this feature until now.

Do you think marking `IEnumerable.AsQueryable()` as unsafe is preferable here? (Note we will still need to suppress the warnings here, since the trimmer doesn't make the connection that the constructor is marked unsafe.)

Reply:
If we leave it where it is - is there a chance that linker will not emit any warning if I don't do any of the "interesting" things. Meaning - if I use simple `(new string[] { "foo" }).AsQueryable().GetEnumerator()` will that warn with the current state of things in this PR?

Reply:
I turned on warnings for the Queryable trimming test we have in the repo:

https://github.com/dotnet/runtime/blob/dacacd0f1e161f9628f1bf8a5ba4f145b466033a/src/libraries/System.Linq.Queryable/tests/TrimmingTests/QueryableTests.cs#L11-L22

Here are the relevant warnings I get:

```
C:\git\runtime\artifacts\bin\trimmingTests\projects\System.Linq.Queryable.TrimmingTests\QueryableTests\win-x64\QueryableTests.cs(18,9): Trim analysis error IL2026: .Program.Main(String[]): Using method 'System.Linq.Expressions.Expression.Property(Expression,MethodInfo)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. The Property metadata or other accessor may be trimmed.. [C:\git\runtime\artifacts\bin\trimmingTests\projects\System.Linq.Queryable.TrimmingTests\QueryableTests\win-x64\project.csproj]

C:\git\runtime\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\UnaryExpression.cs(735,13): Trim analysis error IL2026: System.Linq.Expressions.Expression.Convert(Expression,Type): Using method 'System.Linq.Expressions.Expression.Convert(Expression,Type,MethodInfo)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Creating Expressions requires unreferenced code because the members being referenced by the Expression may be trimmed.. [C:\git\runtime\artifacts\bin\trimmingTests\projects\System.Linq.Queryable.TrimmingTests\QueryableTests\win-x64\project.csproj]
C:\git\runtime\src\libraries\System.Linq.Expressions\src\System\Dynamic\Utils\TypeExtensions.cs(21,13): Trim analysis error IL2070: System.Dynamic.Utils.TypeExtensions.GetAnyStaticMethodValidated(Type,String,Type[]): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Type.GetMethod(String,BindingFlags,Binder,Type[],ParameterModifier[])'. The parameter 'type' of method 'System.Dynamic.Utils.TypeExtensions.GetAnyStaticMethodValidated(Type,String,Type[])' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [C:\git\runtime\artifacts\bin\trimmingTests\projects\System.Linq.Queryable.TrimmingTests\QueryableTests\win-x64\project.csproj]

C:\git\runtime\src\libraries\System.Linq.Queryable\src\System\Linq\EnumerableRewriter.cs(49,21): Trim analysis error IL2026: System.Linq.EnumerableRewriter.VisitMethodCall(MethodCallExpression): Using method 'System.Linq.EnumerableRewriter.FindMethod(Type,String,ReadOnlyCollection<Expression>,Type[])' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Enumerating in-memory collections as IQueryable can require unreferenced code because expressions referencing IQueryable extension methods can get rebound to IEnumerable extension methods. The IEnumerable extension methods could be trimmed causing the application to fail at runtime.. [C:\git\runtime\artifacts\bin\trimmingTests\projects\System.Linq.Queryable.TrimmingTests\QueryableTests\win-x64\project.csproj]
```

Notes:

* The first is because https://github.com/mono/linker/issues/1814
* The middle two are still unsolved for Linq.Expressions, which are being tracked by https://github.com/mono/linker/issues/1821
* The last is this warning.

The difference is the application will get a single warning as opposed to warnings in every place `AsQueryable()` is called.

Reply:
I've changed to to mark `AsQueryable()` as `RequiresUnreferencedCode`. This is the most correct thing to do.

Please take a look.